java/ql/src/experimental/Security/CWE/CWE-200/SensitiveAndroidFileLeak.ql

```diff
@@ -0,0 +1,83 @@
+/**
+ * @name Leaking sensitive Android file
+ * @description Getting file intent from user input without path validation could leak arbitrary
+ *              Android configuration file and sensitive user data.
+ * @kind path-problem
+ * @id java/sensitive-android-file-leak
+ * @problem.severity warning
+ * @tags security
+ *       external/cwe/cwe-200
+ */
+
+import java
+import semmle.code.java.controlflow.Guards
+import AndroidFileIntentSink
+import AndroidFileIntentSource
+import DataFlow::PathGraph
+
+private class StartsWithSanitizer extends DataFlow::BarrierGuard {
+  StartsWithSanitizer() { this.(MethodAccess).getMethod().hasName("startsWith") }
+
+  override predicate checks(Expr e, boolean branch) {
+    e =
+      [
+        this.(MethodAccess).getQualifier(),
+        this.(MethodAccess).getQualifier().(MethodAccess).getQualifier()
+      ] and
+    branch = false
+  }
+}
+
+class AndroidFileLeakConfig extends TaintTracking::Configuration {
+  AndroidFileLeakConfig() { this = "AndroidFileLeakConfig" }
+
+  /**
+   * Holds if `src` is a read of some Intent-typed method argument guarded by a check like
+   * `requestCode == REQUEST_CODE__SELECT_CONTENT_FROM_APPS`, where `requestCode` is the first
+   * argument to `Activity.onActivityResult`.
+   */
+  override predicate isSource(DataFlow::Node src) {
+    exists(
+      AndroidActivityResultInput ai, AndroidFileIntentInput fi, ConditionBlock cb,
+      VarAccess intentVar
+    |
+      cb.getCondition().getAChildExpr().(CompileTimeConstantExpr).getIntValue() =
+        fi.getRequestCode() and
+      cb.getCondition().getAChildExpr() = ai.getRequestCodeVar() and
+      intentVar.getType() instanceof TypeIntent and
+      cb.getBasicBlock() = intentVar.(Argument).getAnEnclosingStmt() and
```

Comment:
Just occurred to me, do you mean to check that intentVar is a parameter in this context, or that it's an argument to some call? That is, do you want the access to x in the context `f(int x) { return x; }` (the `x` in `return x` is a `VarAccess` to a `Parameter`), or `f() { g(x); }` (the `x` in `g(x)` is a `VarAccess` and also an `Argument` because it is passed directly to some callee)?

Reply:
The original design is the latter, which is an argument to some call. Frequently the intent variable is passed to a separate method call e.g. <code>loadOfContentFromApps(data, resultCode)</code>. However, removing this constraint makes the query more generic although the intent variable may appear as multiple sources if it is accessed multiple times in the <code>onActivityResult</code> method. I've changed expected test results to accommodate this change. Please let me know if there is a better way to model this. Thanks.